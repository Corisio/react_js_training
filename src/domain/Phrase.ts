export interface Phrase {
    text: string;
    id: string;
}