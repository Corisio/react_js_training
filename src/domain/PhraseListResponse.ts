export interface PhraseListResponse {
    type: string;
    value: Array<{
        id: number;
        joke: string;
    }>;
}