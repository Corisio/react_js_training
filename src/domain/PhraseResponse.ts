export interface PhraseResponse {
    type: string;
    value: {
        id: number;
        joke: string;
    };
}