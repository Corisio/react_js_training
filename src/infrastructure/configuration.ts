export const API_URL = 'http://api.icndb.com/jokes/random/';
export const API_MOST_IMPORTANT_URL = 'http://api.icndb.com/jokes/52';