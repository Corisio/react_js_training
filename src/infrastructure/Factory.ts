import axios, {AxiosInstance} from 'axios';
import PhraseService from '../service/PhraseService';

const httpClient  : AxiosInstance = axios.create();
export const phraseService : PhraseService = new PhraseService(httpClient);