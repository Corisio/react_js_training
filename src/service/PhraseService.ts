import {AxiosInstance, AxiosResponse} from 'axios';
import { Phrase } from '../domain/Phrase';
import { PhraseListResponse } from '../domain/PhraseListResponse';
import { PhraseResponse} from '../domain/PhraseResponse';
import { API_URL, API_MOST_IMPORTANT_URL} from '../infrastructure/configuration';

class PhraseService {
    private httpClient:AxiosInstance;

    constructor(httpClient:AxiosInstance){
        this.httpClient = httpClient;
    }

    async getRandomPhrases(number: number): Promise<Array<Phrase>>{
        const response: AxiosResponse<PhraseListResponse> = await this.httpClient.get(API_URL + number.toString());
        
        return parsePhraseList(response.data);
    }

    async getMostImportantPhrase(): Promise<Phrase>{
        const response: AxiosResponse<PhraseResponse> = await this.httpClient.get(API_MOST_IMPORTANT_URL);
        
        return parsePhrase(response.data);
    }
}

function parsePhraseList(phraseResponse: PhraseListResponse) : Array<Phrase>{
    const result: Array<Phrase> = phraseResponse.value.map(
        phraseObject => ({id: phraseObject.id.toString(), text: phraseObject.joke}));
    
    return result;
}

function parsePhrase(phraseResponse: PhraseResponse) : Phrase{
    const result: Phrase = {id: phraseResponse.value.id.toString(), text: phraseResponse.value.joke};
    
    return result;
}

export default PhraseService;