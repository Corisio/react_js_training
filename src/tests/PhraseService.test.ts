import axios, {AxiosInstance, AxiosResponse} from 'axios';
import MockAdapter from 'axios-mock-adapter';

import PhraseService from '../service/PhraseService';
import { Phrase } from '../domain/Phrase';
import { RandomPhrases } from './fixtures/services/randomPhrases';
import { API_URL } from '../infrastructure/configuration';

const httpClient = axios.create();
const axiosMocked = new MockAdapter(httpClient);

describe('PhraseService', function(){
    it('get phrases', async function(){
        const phraseCount: number = 5;
        const jokes = RandomPhrases.slice(0, phraseCount).map(phrase => ({"id": phrase.id, "joke": phrase.text}));
        axiosMocked.onGet(API_URL + phraseCount.toString()).reply
            (200, { "type": "success", "value": jokes});
        const phraseService: PhraseService = new PhraseService(httpClient);

        const actualPhrases: Array<Phrase> = await phraseService.getRandomPhrases(phraseCount);

        expect(actualPhrases).toEqual(RandomPhrases);
        expect(actualPhrases.length).toEqual(phraseCount);
        expect(actualPhrases).toMatchSnapshot();
    });
});