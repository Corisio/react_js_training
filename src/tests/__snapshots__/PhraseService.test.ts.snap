// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhraseService get phrases 1`] = `
Array [
  Object {
    "id": "39",
    "text": "Chuck Norris can win at solitaire with only 18 cards.",
  },
  Object {
    "id": "40",
    "text": "Chuck Norris? roundhouse kick is so powerful, it can be seen from outer space by the naked eye.",
  },
  Object {
    "id": "41",
    "text": "Using his trademark roundhouse kick, Chuck Norris once made a fieldgoal in RJ Stadium in Tampa Bay from the 50 yard line of Qualcomm stadium in San Diego.",
  },
  Object {
    "id": "42",
    "text": "When Chuck Norris throws exceptions, it's across the room.",
  },
  Object {
    "id": "43",
    "text": "Chuck Norris uses a night light. Not because Chuck Norris is afraid of the dark, but the dark is afraid of Chuck Norris.",
  },
]
`;
